### PR TITLE
Use ResourceLoggerService and ResourceNotification service for DCP based resource changes

### LINKED ISSUE #2548: Dashboard is appending the path from "launchUrl" on launch profile to endpoints that weren't added from the launch profile
If I add a custom HTTP endpoint to a project resource in addition to the default endpoints discovered from the launch profile, and the launch profile has a path set on the `"launchUrl"` property, the dashboard includes that path on the hyperlink rendered for the custom endpoints. The path from the launch profile should only be appended to the endpoint addresses from the launch profile.

```csharp
var apiService = builder.AddProject<Projects.HealthChecksUI_ApiService>("apiservice")
    .WithHttpEndpoint(name: "healthchecks");
```

image

### PR DESCRIPTION
**Pull request by davidfowl**
The goal here is to hand a single way to handle resource notification and logs by using the interface we introduced for custom resources everywhere. The dashboard data model now depends on these services and nothing DCP specific. The god class `ApplicationExecutor` now watches for changes after initializing services.

- Moved all of the DCP logic calling into k8s APIs into the ApplicationExecutor.
- Added an overload of WatchAsync that watches all resources. This is what the dashboard uses.
- Watch logs as soon as they become available an push them through the
  ResourceLoggerService API. This makes them avaialble to anyone looking
  for logs.
- Support sending updates based on IResource and a unique resource id. This is how we can support unique snapshots for replicas.
- Introduce a "Hidden" state to allow hiding resources from the dashboard views. Use this to hide app model resources after DCP has loaded the resource.
- Watching resource state changes are global. This makes it possible to listen to changes to replicas since they aren't directly represented in the model, but they are still associated with an IResource.
- Publish the initial state of any resource on start.

Fixes https://github.com/dotnet/aspire/issues/2548

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2731)

### CODE REVIEW

File: src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs [hunk 1]
```diff
@@ -30,13 +35,28 @@ public sealed record CustomResourceSnapshot
...
-    public ImmutableArray<(string Name, string Value)> EnvironmentVariables { get; init; } = [];
+    public ImmutableArray<(string Name, string Value, bool IsFromSpec)> EnvironmentVariables { get; init; } = [];
 
     /// <summary>
     /// The URLs that should show up in the dashboard for this resource.
     /// </summary>
     public ImmutableArray<(string Name, string Url)> Urls { get; init; } = [];
+
+    /// <summary>
+    /// The endpoints that should show up in the dashboard for this resource.
```
Comment (karolz-ms) [inline src/Aspire.Hosting/ApplicationModel/CustomResourceSnapshot.cs:54]: Needs updating

File: src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs [hunk 2]
```diff
@@ -50,6 +58,19 @@ public void Complete(IResource resource)
             logger.Complete();
         }
     }
+
+    /// <summary>
+    /// Completes the log stream for the resource.
+    /// </summary>
+    /// <param name="name">The name of the resource.</param>
+    public void Complete(string name)
```
Comment (karolz-ms) [inline src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs:66]: Should it be an error, or at least should it be possible to find out that one tried to "complete" a log stream for a resource that does not exist?

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 3]
```diff
@@ -80,14 +91,52 @@ public Task PublishUpdateAsync(IResource resource, Func<CustomResourceSnapshot,
...
+            }
+
+            var channel = Channel.CreateUnbounded<ResourceEvent>();
+
+            void WriteToChannel(ResourceEvent resourceEvent)
+            {
+                channel.Writer.TryWrite(resourceEvent);
+            }
+
+            resourceNotificationService.OnResourceUpdated += WriteToChannel;
```
Comment (karolz-ms) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:123]: Is this thread-safe?

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs [hunk 4]
```diff
@@ -164,3 +213,21 @@ await foreach (var item in channel.Reader.ReadAllAsync(cancellationToken))
         }
     }
 }
+
+/// <summary>
+/// Represents a change in the state of a resource.
+/// </summary>
+/// <param name="resource">The resource associated with the event. </param>
+/// <param name="snapshot">The snapshot of the resource state.</param>
+public class ResourceEvent(IResource resource, CustomResourceSnapshot snapshot)
```
Comment (karolz-ms) [inline src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:222]: Should this be a "record"?

File: src/Aspire.Hosting/Dashboard/DashboardServiceData.cs [hunk 5]
```diff
@@ -17,22 +16,70 @@ internal sealed class DashboardServiceData : IAsyncDisposable
...
+                    Uid = name,
+                    CreationTimeStamp = snapshot.CreationTimeStamp ?? creationTimestamp,
+                    Name = name,
+                    DisplayName = resource.GetReplicaCount() is 1 ? resource.Name : name,
+                    Endpoints = endpoints,
+                    Environment = environmentVariables,
+                    ExitCode = snapshot.ExitCode,
+                    ExpectedEndpointsCount = endpoints.Length,
+                    Services = services,
+                    State = snapshot.State ?? "Running"
```
Comment (karolz-ms) [inline src/Aspire.Hosting/Dashboard/DashboardServiceData.cs:62]: Should we really tell people that everything is A-OK if we do not know?

File: src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs:66]: I'll return a bool.

File: src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs:123]: Yep

File: src/Aspire.Hosting/Dashboard/DashboardServiceData.cs (hunk 5 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Dashboard/DashboardServiceData.cs:62]: I think not. Now we have this https://github.com/dotnet/aspire/commit/e86f9fd285fa32fc59d1151e9cfc92b9959ea9e4, so I'll remove the fallback.

**Review by mitchdenny — APPROVED** [review]

**Pull request merged**